### PR TITLE
element-desktop: add darwin support

### PR DESCRIPTION
**Pull request by eliandoran**
## Description of changes

Unpacks the original macOS binary from https://packages.element.io/desktop/update/macos/index.html.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/element/element-desktop.nix [hunk 1]
```diff
@@ -17,139 +17,172 @@
...
+    };
+
+    meta = with lib; metaCommon // {
+      maintainers = teams.matrix.members;
+      mainProgram = "element-desktop";
+    };
+  });
+
+  darwin = stdenv.mkDerivation (finalAttrs: builtins.removeAttrs pinData [ "hashes" ] // {
+    pname = pname;
```
Comment (Aleksanaa) [inline pkgs/applications/networking/instant-messengers/element/element-desktop.nix:166]:
```suggestion
    inherit pname;
```

File: pkgs/applications/networking/instant-messengers/element/element-desktop.nix [hunk 2]
```diff
@@ -17,139 +17,172 @@
...
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p "$out/Applications/Element.app"
+      cp -R . "$out/Applications/Element.app"
+      runHook postInstall
+    '';
+
+    meta = with lib; metaCommon // {
+      maintainers = with maintainers; [ eliandoran ];
```
Comment (Aleksanaa) [inline pkgs/applications/networking/instant-messengers/element/element-desktop.nix:183]:
```suggestion
      maintainers = with maintainers; [ eliandoran ];
      sourceProvenance = with sourceTypes; [ binaryNativeCode ];
```

**Comment by leona-ya** [comment]
Do you have problems with element-desktop on darwin? For me it already works fine with the current from-source variant on aarch64-darwin.

**Comment by eliandoran** [comment]
@leona-ya , good point...

I had this issue when trying to test on nixos 23.11:

```
error: Package ‘element-desktop-1.11.59’ in /Users/elian/Projects/nixpkgs/pkgs/applications/networking/instant-messengers/element/element-desktop.nix:147 is not available on the requested hostPlatform:
         hostPlatform.config = "x86_64-apple-darwin"
         package.meta.platforms = [
           "aarch64-linux"
           "armv5tel-linux"
           "armv6l-linux"
           "armv7a-linux"
           "armv7l-linux"
           "i686-linux"
           "loongarch64-linux"
           "m68k-linux"
           "microblaze-linux"
           "microblazeel-linux"
           "mips-linux"
           "mips64-linux"
           "mips64el-linux"
           "mipsel-linux"
           "powerpc64-linux"
           "powerpc64le-linux"
           "riscv32-linux"
           "riscv64-linux"
           "s390-linux"
           "s390x-linux"
           "x86_64-linux"
         ]
         package.meta.badPlatforms = [ ]
       , refusing to evaluate.
```

Seems that `element-desktop` is building just fine on `master`, but for some reason I didn't bother to check. 🤦🏻 
Thanks!

**Pull request closed without merging**